File: Youtube_java/user/PremiumUser.java
package user;
import java.util.*;
import java.io.*;
import video.Video;
import user.User;
import idGenerator.*;

public class PremiumUser extends User {
    private static final long serialVersionUID = 1L;
    
    public String paymentSubcript;

    public PremiumUser() {
        super();
    }

    public PremiumUser(String name, String gender, int numtel, String email, String password) {
        super();
        this.userId = IdGenerator.generateUserId();
        this.nameUser = name;
        this.gender = gender;
        this.numTel = numtel;
        this.email = email;
        this.password = password;
    }

    public String getPaymentSubcript() {
        return paymentSubcript;
    }

    public void setPaymentSubcript(String paymentSubcript) {
        this.paymentSubcript = paymentSubcript;
    }

    public static PremiumUser createPremiumUser(String name, String gender, int numtel, String email, String password, String option) {
        PremiumUser premiumUser = new PremiumUser(name, gender, numtel, email, password);
        premiumUser.paymentSubcript = option;
        return premiumUser;
    }

    // Souscrire à une option
    public PremiumUser subscriptionPayment(Scanner sc) {
        System.out.println("\n--- OPTIONS DE SOUSCRIPTION PREMIUM ---");
        System.out.println("1....................OPTION 1 : 2,05$/ mois .....................");
        System.out.println("2....................OPTION 2 : 5,57$/ mois .....................");
        System.out.println("3....................OPTION 3 : 12,89$/ mois ....................");
        System.out.print("Quelle option desirez vos souscrire ?..");

        if (!sc.hasNextInt()) {
            System.out.println("Choix invalide. Veuillez entrer un nombre.");
            sc.nextLine();
            return this;
        }

        int choix = sc.nextInt();
        sc.nextLine();

        String optionDescription = "";

        switch (choix) {
            case 1:
                optionDescription = "OPTION 1 : 2,05$/mois";
                System.out.println("VOUS SOUSCRIVREZ BIEN À L'" + optionDescription + "......... ");
                System.out.print("Inserez le numero de votre carte bancaire :");
                String cardNumber = sc.nextLine();
                this.setPaymentSubcript(optionDescription);
                System.out.println(" PAIEMENT ET OPERATION REUSSIS");
                break;
            case 2:
                optionDescription = "OPTION 2 : 5,57$/mois";
                System.out.println("VOUS SOUSCRIVREZ BIEN À L'" + optionDescription + "......... ");
                String cardNber = sc.nextLine();
                this.setPaymentSubcript(optionDescription);
                System.out.println(" PAIEMENT ET OPERATION REUSSIS");
                break;
            case 3:
                optionDescription = "OPTION 3 : 12,89$/mois";
                System.out.println("VOUS SOUSCRIVREZ BIEN À L'" + optionDescription + "......... ");
                String cardNum = sc.nextLine();
                this.setPaymentSubcript(optionDescription);
                System.out.println(" PAIEMENT ET OPERATION REUSSIS");
                break;
            default:
                System.out.println("Choix d'option invalide. Souscription non effectuée.");
                break;
        }
        
        saveToFile(); // Sauvegarder après souscription
        return this;
    }

    // Profil du premium user
    public void displayProfilePremiumUser(PremiumUser p) {
        System.out.println("\t \t ***************** INFORMATION DE L'UTILISATEUR **********************\n");
        System.out.println("ID Utilisateur :" + p.getUserId());
        System.out.println("Nom :" + p.getNameUser());
        System.out.println("Genre :" + p.getGender());
        System.out.println("Numero de Telephone :" + p.getNumTel());
        System.out.println("Email :" + p.getEmail());
        System.out.println("Mot de Passe :" + p.getPassword());
        System.out.println("Option de souscription:" + p.getPaymentSubcript());
        System.out.println("Nombre de vidéos créées: " + p.userVideos.size());
        System.out.println("Nombre de likes: " + p.likedVideos.size());
        System.out.println("Nombre de commentaires: " + p.userComments.size());
    }
    
    @Override
    public void saveToFile() {
        try (PrintWriter writer = new PrintWriter(new FileWriter(userId + ".txt"))) {
            // Informations de base avec type Premium
            writer.println("USERTYPE:PREMIUM");
            writer.println("USERINFO:" + userId + "|" + nameUser + "|" + gender + "|" + 
                          numTel + "|" + email + "|" + password + "|" + 
                          (paymentSubcript != null ? paymentSubcript : ""));
            
            // Vidéos créées
            writer.println("VIDEOS:" + userVideos.size());
            for (Video video : userVideos) {
                writer.println(video.serialize());
            }
            
            // Vidéos likées
            writer.println("LIKES:" + likedVideos.size());
            for (String videoId : likedVideos) {
                writer.println(videoId);
            }
            
            // Commentaires
            writer.println("COMMENTS:" + userComments.size());
            for (Comments comment : userComments) {
                writer.println(comment.serialize());
            }
            
            System.out.println("Données utilisateur premium sauvegardées dans " + userId + ".txt");
        } catch (IOException e) {
            System.err.println("Erreur lors de la sauvegarde du fichier utilisateur: " + e.getMessage());
        }
    }
    
    public static PremiumUser loadPremiumFromFile(String userId) {
        try (BufferedReader reader = new BufferedReader(new FileReader(userId + ".txt"))) {
            PremiumUser user = new PremiumUser();
            String line;
            
            // Vérifier le type
            if ((line = reader.readLine()) != null && !line.equals("USERTYPE:PREMIUM")) {
                return null;
            }
            
            // Lire les informations de base
            if ((line = reader.readLine()) != null && line.startsWith("USERINFO:")) {
                String[] parts = line.substring(9).split("\\|");
                user.userId = parts[0];
                user.nameUser = parts[1];
                user.gender = parts[2];
                user.numTel = Integer.parseInt(parts[3]);
                user.email = parts[4];
                user.password = parts[5];
                if (parts.length > 6) {
                    user.paymentSubcript = parts[6];
                }
            }
            
            // Lire les vidéos
            if ((line = reader.readLine()) != null && line.startsWith("VIDEOS:")) {
                int count = Integer.parseInt(line.substring(7));
                for (int i = 0; i < count; i++) {
                    line = reader.readLine();
                    if (line != null) {
                        user.userVideos.add(Video.deserialize(line));
                    }
                }
            }
            
            // Lire les likes
            if ((line = reader.readLine()) != null && line.startsWith("LIKES:")) {
                int count = Integer.parseInt(line.substring(6));
                for (int i = 0; i < count; i++) {
                    line = reader.readLine();
                    if (line != null) {
                        user.likedVideos.add(line);
                    }
                }
            }
            
            // Lire les commentaires
            if ((line = reader.readLine()) != null && line.startsWith("COMMENTS:")) {
                int count = Integer.parseInt(line.substring(9));
                for (int i = 0; i < count; i++) {
                    line = reader.readLine();
                    if (line != null) {
                        user.userComments.add(Comments.deserialize(line));
                    }
                }
            }
            
            return user;
        } catch (IOException e) {
            System.err.println("Erreur lors du chargement du fichier utilisateur premium: " + e.getMessage());
            return null;
        }
    }
}

File: Youtube_python/premium_user.py
from user import User
from id_generator import IdGenerator
import os

class PremiumUser(User):
    def __init__(self):
        super().__init__()
        self.payment_subscript = None
    
    @staticmethod
    def create_premium_user(name, gender, num_tel, email, password, option):
        premium_user = PremiumUser()
        premium_user.name_user = name
        premium_user.gender = gender
        premium_user.num_tel = num_tel
        premium_user.email = email
        premium_user.password = password
        premium_user.payment_subscript = option
        return premium_user
    
    def subscription_payment(self):
        print("\n--- OPTIONS DE SOUSCRIPTION PREMIUM ---")
        print("1....................OPTION 1 : 2,05$/ mois ...................")
        print("2....................OPTION 2 : 5,57$/ mois ...................")
        print("3....................OPTION 3 : 12,89$/ mois ..................")
        
        try:
            choix = int(input("Quelle option desirez vous souscrire ?.. "))
        except ValueError:
            print("Choix invalide. Veuillez entrer un nombre.")
            return self
        
        option_description = ""
        
        if choix == 1:
            option_description = "OPTION 1 : 2,05$/mois"
            print(f"VOUS SOUSCRIVREZ BIEN À L'{option_description}......... ")
            card_number = input("Inserez le numero de votre carte bancaire :")
            self.payment_subscript = option_description
            print(" PAIEMENT ET OPERATION REUSSIS")
        elif choix == 2:
            option_description = "OPTION 2 : 5,57$/mois"
            print(f"VOUS SOUSCRIVREZ BIEN À L'{option_description}......... ")
            card_number = input("Inserez le numero de votre carte bancaire :")
            self.payment_subscript = option_description
            print(" PAIEMENT ET OPERATION REUSSIS")
        elif choix == 3:
            option_description = "OPTION 3 : 12,89$/mois"
            print(f"VOUS SOUSCRIVREZ BIEN À L'{option_description}......... ")
            card_number = input("Inserez le numero de votre carte bancaire :")
            self.payment_subscript = option_description
            print(" PAIEMENT ET OPERATION REUSSIS")
        else:
            print("Choix d'option invalide. Souscription non effectuée.")
        
        self.save_to_file()
        return self
    
    def display_profile_premium_user(self):
        print("\t \t ***************** INFORMATION DE L'UTILISATEUR **********************\n")
        print(f"ID Utilisateur : {self.user_id}")
        print(f"Nom : {self.name_user}")
        print(f"Genre : {self.gender}")
        print(f"Numero de Telephone : {self.num_tel}")
        print(f"Email : {self.email}")
        print(f"Mot de Passe : {self.password}")
        print(f"Option de souscription: {self.payment_subscript}")
        print(f"Nombre de vidéos créées: {len(self.user_videos)}")
        print(f"Nombre de likes: {len(self.liked_videos)}")
        print(f"Nombre de commentaires: {len(self.user_comments)}")
    
    def save_to_file(self):
        try:
            with open(f"{self.user_id}.txt", 'w', encoding='utf-8') as f:
                # Type Premium
                f.write("USERTYPE:PREMIUM\n")
                f.write(f"USERINFO:{self.user_id}|{self.name_user}|{self.gender}|{self.num_tel}|{self.email}|{self.password}|{self.payment_subscript or ''}\n")
                
                # Vidéos
                f.write(f"VIDEOS:{len(self.user_videos)}\n")
                for video in self.user_videos:
                    f.write(f"{video.serialize()}\n")
                
                # Likes
                f.write(f"LIKES:{len(self.liked_videos)}\n")
                for video_id in self.liked_videos:
                    f.write(f"{video_id}\n")
                
                # Commentaires
                f.write(f"COMMENTS:{len(self.user_comments)}\n")
                for comment in self.user_comments:
                    f.write(f"{comment.serialize()}\n")
            
            print(f"Données utilisateur premium sauvegardées dans {self.user_id}.txt")
        except Exception as e:
            print(f"Erreur lors de la sauvegarde du fichier utilisateur: {e}")
    
    @staticmethod
    def load_premium_from_file(user_id):
        # Import ici pour éviter l'importation circulaire
        from video import Video
        from comments import Comments
        
        filename = f"{user_id}.txt"
        if not os.path.exists(filename):
            return None
        
        try:
            with open(filename, 'r', encoding='utf-8') as f:
                line = f.readline().strip()
                if line != "USERTYPE:PREMIUM":
                    return None
                
                line = f.readline().strip()
                if not line.startswith("USERINFO:"):
                    return None
                
                parts = line[9:].split('|')
                premium_user = PremiumUser()
                premium_user.user_id = parts[0]
                premium_user.name_user = parts[1]
                premium_user.gender = parts[2]
                premium_user.num_tel = int(parts[3])
                premium_user.email = parts[4]
                premium_user.password = parts[5]
                if len(parts) > 6:
                    premium_user.payment_subscript = parts[6]
                
                # Lire vidéos
                line = f.readline().strip()
                if line.startswith("VIDEOS:"):
                    count = int(line[7:])
                    for _ in range(count):
                        line = f.readline().strip()
                        if line:
                            premium_user.user_videos.append(Video.deserialize(line))
                
                # Lire likes
                line = f.readline().strip()
                if line.startswith("LIKES:"):
                    count = int(line[6:])
                    for _ in range(count):
                        line = f.readline().strip()
                        if line:
                            premium_user.liked_videos.append(line)
                
                # Lire commentaires
                line = f.readline().strip()
                if line.startswith("COMMENTS:"):
                    count = int(line[9:])
                    for _ in range(count):
                        line = f.readline().strip()
                        if line:
                            premium_user.user_comments.append(Comments.deserialize(line))
                
                return premium_user
        except Exception as e:
            print(f"Erreur lors du chargement du fichier utilisateur premium: {e}")
            return None
